### PR TITLE
Prototype implementing DataFusion functions / operators using arrow-udf liibrary

### LINKED ISSUE #11413: Prototype implementing DataFusion functions / operators using `arrow-udf` liibrary
### Is your feature request related to a problem or challenge?

Related to the discussion on https://github.com/apache/datafusion/discussions/11192 with @Xuanwo

RisingWave has a library for automatically creating vectorized implementations of functions (e.g. that operate on arrow arrays) from scalar implementations

The library is here: https://github.com/risingwavelabs/arrow-udf

A blog post describing it is here: https://risingwave.com/blog/simplifying-sql-function-implementation-with-rust-procedural-macro/

DataFusion uses macros to do something similar in binary.rs but they are pretty hard to read / understand in my opinon: https://github.com/apache/datafusion/blob/7a23ea9bce32dc8ae195caa8ca052673031c06c9/datafusion/physical-expr/src/expressions/binary.rs#L118-L130

One main benefit I can see to switching to https://github.com/risingwavelabs/arrow-udf is that we could then extend arrow-udf to support Dictionary and StringView and maybe other types to generate fast kernels for multiple different array layouts. 

### Describe the solution you'd like

I think it would be great if someone could evaluate the feasibility of using the macros in https://github.com/risingwavelabs/arrow-udf to implement Datafusion's operations (and maybe eventually functions etc)

### Describe alternatives you've considered

I suggest a POC that picks one or two functions (maybe string equality or regexp_match or something) and tries to use `arrow-udf`s function macro instead. 

Here is an example of how to use it: https://docs.rs/arrow-udf/0.3.0/arrow_udf/

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xinlifoobar**
## Which issue does this PR close?

Closes #11413

### CODE REVIEW

File: datafusion/functions/src/udf.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+use arrow::{
+    array::{Array, ArrayRef, RecordBatch},
+    datatypes::{Field, Schema, SchemaRef},
+};
+use arrow_udf::{function, sig::REGISTRY};
+use datafusion_common::{internal_err, DataFusionError, Result};
+use datafusion_expr::ColumnarValue;
+// use arrow_string::predicate::Predicate;
+
+#[function("eq(boolean, boolean) -> boolean")]
```
Comment (alamb) [inline datafusion/functions/src/udf.rs:29]: this is pretty neat

File: datafusion/functions/src/udf.rs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+//     Predicate::like(rhs).unwrap().matches(lhs);
+// }
+
+#[function("concat(string, string) -> string")]
+#[function("concat(largestring, largestring) -> largestring")]
+fn concat(lhs: &str, rhs: &str) -> String {
+    format!("{}{}", lhs, rhs)
+}
+
+pub fn apply_udf(
```
Comment (alamb) [inline datafusion/functions/src/udf.rs:59]: This is the code that adapts `ColumnarValue` to being able to call the `arrow-udf` based functions via a record batch, right?

**Review by alamb — COMMENTED** [review]
This is very cool @xinlifoobar  -- thank you

**Comment by alamb** [comment]
I will leave comments on https://github.com/apache/datafusion/issues/11413

**Comment by Xuanwo** [comment]
Hello, everyone. Exciting news: arrow-udf now belongs to a new organization called [`arrow-udf`](https://github.com/arrow-udf). Please let me know if there is anything preventing us from moving forward. I'm happy to help address it.

**Pull request closed without merging**